src/components/ColumnPicker/useColumns.ts

```diff
@@ -0,0 +1,102 @@
+import useStateFromProps from "@dashboard/hooks/useStateFromProps";
+import { addAtIndex, removeAtIndex } from "@dashboard/utils/lists";
+import { GridColumn } from "@glideapps/glide-data-grid";
+import React from "react";
+
+import { AvailableColumn } from "../Datagrid/types";
+
+export interface ColumnCategory {
+  name: string;
+  prefix: string;
+  availableNodes: AvailableColumn[];
+  selectedNodes: AvailableColumn[];
+  onSearch?: (query: string) => void;
+  hasNextPage: boolean;
+  hasPreviousPage: boolean;
+  onNextPage: (query: string) => void;
+  onPreviousPage: (query: string) => void;
+}
+
+export interface UseColumnsProps {
+  staticColumns: AvailableColumn[];
+  columnCategories?: ColumnCategory[];
+  selectedColumns: string[];
+  onSave: (columns: string[]) => void;
+  columnPickerSettings: string[];
+  setDynamicColumnSettings: (cols: string[]) => void;
+}
+
+export const useColumns = ({
+  staticColumns,
+  selectedColumns,
+  columnCategories,
+  onSave,
+  columnPickerSettings,
+  setDynamicColumnSettings,
+}: UseColumnsProps) => {
+  const dynamicColumns: AvailableColumn[] = React.useMemo(
+    () =>
+      columnCategories.reduce((columns: AvailableColumn[], category) => {
+        category.selectedNodes.forEach(column => {
+          if (columnPickerSettings.includes(column.id)) {
+            columns.push(column);
+          }
+        });
+        return columns;
+      }, [] as AvailableColumn[]),
+    [columnCategories, columnPickerSettings],
+  );
+
+  const initialColumnsState = React.useMemo(
+    () =>
+      [...staticColumns, ...dynamicColumns].filter(
+        column => selectedColumns.includes(column.id) || column.id === "empty",
+      ),
+    [dynamicColumns, staticColumns, selectedColumns],
+  );
+
+  const [visibleColumns, setVisibleColumns] =
+    useStateFromProps(initialColumnsState);
```

Comment:
suggestion: bro this is deprecated

Reply:
We deprecated it because it was used excessively in the dashboard, however in this case this is exactly what is needed - useState with state update based on deep comparisons. Reference update (useStateUpdate) will lead to infinite loop there and shallow comparisons won't work too because changes are deep in the object (and hook like this is not present in the dashboard too). We might think about undoing the depreciation on this hook, just with warning about its usage, because it's heavy artillery and usually there are better solutions.